tests/boards/intel_adsp/hda/src/smoke.c

```diff
@@ -0,0 +1,189 @@
+					    STREAM_ID, IPC_TIMEOUT));
+	printk("host start: "); cavs_hda_dbg("host_in", HDA_HOST_IN_BASE, STREAM_ID);
+
+	for (uint32_t i = 0; i < TRANSFER_COUNT; i++) {
+		cavs_hda_commit(HDA_HOST_IN_BASE, STREAM_ID, FIFO_SIZE);
+		printk("dsp inc_pos: "); cavs_hda_dbg("host_in", HDA_HOST_IN_BASE, STREAM_ID);
+
+		WAIT_FOR(cavs_hda_wp_rp_eq(HDA_HOST_IN_BASE, STREAM_ID));
+		printk("dsp wp_rp_eq: "); cavs_hda_dbg("host_in", HDA_HOST_IN_BASE, STREAM_ID);
+
+		k_sleep(K_MSEC(10));
+
+		last_msg_cnt = msg_cnt;
+		WAIT_FOR(cavs_ipc_send_message_sync(CAVS_HOST_DEV, IPCCMD_HDA_VALIDATE,
+						    STREAM_ID, IPC_TIMEOUT));
+		WAIT_FOR(msg_cnt > last_msg_cnt);
+		zassert_true(msg_res == 1,
+			     "Expected data validation to be true from Host");
+	}
+
+	WAIT_FOR(cavs_ipc_send_message_sync(CAVS_HOST_DEV, IPCCMD_HDA_RESET,
+					    STREAM_ID, IPC_TIMEOUT));
+	cavs_hda_disable(HDA_HOST_IN_BASE, STREAM_ID);
+}
+
+void test_hda_host_out_smoke(void)
+{
+	int res;
+	bool is_ramp;
+
+	printk("smoke testing hda with fifo buffer at address %p, size %d\n",
+	       hda_fifo, FIFO_SIZE);
+
+	cavs_ipc_set_message_handler(CAVS_HOST_DEV, ipc_message, NULL);
+	cavs_ipc_set_done_handler(CAVS_HOST_DEV, ipc_done, NULL);
+
+	printk("Using buffer of size %d at addr %p\n", FIFO_SIZE, hda_fifo);
+
+	cavs_hda_init(HDA_HOST_OUT_BASE, STREAM_ID);
+	printk("dsp init: "); cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+
+	WAIT_FOR(cavs_ipc_send_message_sync(CAVS_HOST_DEV, IPCCMD_HDA_RESET,
+					    (STREAM_ID + 7), IPC_TIMEOUT));
+	printk("host reset: "); cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+
+	WAIT_FOR(cavs_ipc_send_message_sync(CAVS_HOST_DEV, IPCCMD_HDA_CONFIG,
+					    (STREAM_ID + 7) | (FIFO_SIZE << 8),
+					    IPC_TIMEOUT));
+	printk("host config: "); cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+
+	res = cavs_hda_set_buffer(HDA_HOST_OUT_BASE, STREAM_ID, hda_fifo, FIFO_SIZE);
+	printk("dsp set buffer: "); cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+	zassert_ok(res, "Expected set buffer to succeed");
+
+	k_sleep(K_MSEC(10));
+
+
+	for (uint32_t i = 0; i < TRANSFER_COUNT; i++) {
+
+		WAIT_FOR(cavs_ipc_send_message_sync(CAVS_HOST_DEV, IPCCMD_HDA_SEND,
+						    (STREAM_ID + 7) | (FIFO_SIZE << 8),
+						    IPC_TIMEOUT));
+		printk("host send: ");
+		cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+
+		cavs_hda_enable(HDA_HOST_OUT_BASE, STREAM_ID);
+		printk("dsp enable: ");
+		cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+
+		WAIT_FOR(cavs_ipc_send_message_sync(CAVS_HOST_DEV, IPCCMD_HDA_START,
+						    (STREAM_ID + 7), IPC_TIMEOUT));
+		printk("host start: ");
+		cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+
+		WAIT_FOR(cavs_hda_buf_full(HDA_HOST_OUT_BASE, STREAM_ID));
+		printk("dsp wait for full: ");
+		cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+
+		/* The buffer is in the cached address range and must be invalidated
+		 * prior to reading.
```

Comment:
Uh... no it isn't.  Or if it is that's a linkage bug.  Static data needs to be uncached always.  You get caching on the stack, and on symbols with an explicit __incoherent tag applied (**and** where you've taken care to make sure they're correctly aligned and padded and won't overlap a cache line with anyone else).  That's the core metaphor behind KERNEL_COHERENCE.  You can rely on everything except local variables being coherent.

Reply:
More leftover junk here, noted for the cleanup round

Reply:
Presumably I should be able to zassert_true(arch_mem_coherent(hda_fifo), "mem is coherent"); and that fails here for me

Reply:
It's because I had set MP_NUM_CPUS=1 in my test prj.conf trying to reduce what might be going wrong while debugging, this turns off the KERNEL_COHERENCE which is still probably wrong especially given the way HDA works.

Reply:
i've added tests to cover the NUM_CPUS=1 or the platform default to deal with this, as well as wrapping all the cache flush/invalidate bits in ifdefry